### ISSUE TITLE
VariableDeclarationUsageDistanceCheck doesn't handle class definition, function definition properly

### ISSUE THREAD

**Issue by jxr98** [checkstyle/checkstyle#13011]
I have read check documentation: https://checkstyle.org/config_coding.html#VariableDeclarationUsageDistance
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
```

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
          <property name="id" value="Custom"/>
          <property name="validateBetweenScopes" value="true"/>
          <property name="allowedDistance" value="1"/>
        </module> 
    </module>
    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>
```

```
/var/tmp $ cat Test.java
```

```
public class Test {

    void nothing() {
    }

    class Parent {
        void mm() {
        }
    }

public void method2() {
        Integer a = 5;  // violation 'Distance .* is 4.'
        class AClass extends Parent {
            int i;

            public AClass(int i) {
                this.i = i;
            }

            @Override
            void mm() {
                if (a >= 0) {
                    System.out.println("test");
                }
            }
        }
    }
}
```

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.3-all.jar -c config.xml Test.java
```
```
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-X.XX-all.jar -c config.xml Test.java
```
```
Starting audit...
[WARN] /home/jxr/Documents/checkstyle/Test.java:12:9: Distance between variable 'a' declaration and its first usage is 4, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [Custom]
Audit done.
```

---

**Describe what you expect in detail.**
No error because from my point of view, class definition(extends), variable definition and function definition are not counted towards distance. Thus, the distance between variable a's definiton and usage is 1.

---

**Comment by jxr98** [comment]
The following example uses the same config file as above.
`/var/tmp $ cat Test1.java`
```
public class Test1 {
    public interface Ani {
        public void method4(Object o);
    }

    public void method5() {
        // false positive, the distance is 1
        Object a = new Object(); // violation 'Distance .* is 4.'
        class AClass implements Ani {
            @Override
            public void method4(Object o) {
                if (a.toString().isEmpty()) {
                    System.out.println(o.toString());
                }
            }
        }
    }
}
```

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.3-all.jar -c config.xml Test1.java
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-X.XX-all.jar -c config.xml Test1.java
```

```
Starting audit...
[WARN] /var/tmp/Test3.java:8:9: Distance between variable 'a' declaration and its first usage is 4, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [Custom]
Audit done.
```

**Comment by jxr98** [comment]
The following example uses the same config file as above.
`/var/tmp $ cat Test2.java`
```
public class Test2 {
    public void method7() {
        Integer t = 5;
        nothing();
        System.out.println();
        class BClass extends Parent {
            @Override
            void mm() {
                System.out.println(t);
            }
        }
    }
}
```

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.3-all.jar -c config.xml Test2.java
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-X.XX-all.jar -c config.xml Test2.java
```

```
Starting audit...
Audit done.
```
---

**Describe what you expect in detail.**
There should be a warning because `nothing()` and `System.out.println()` should be counted towards distance. The distance between definition and usage is 3.

---

**Comment by jxr98** [comment]
The following example uses the same config file as above.
`/var/tmp $ cat Test2.java`
```
public class Test3 {
    class Parent {
        void mm() {
        }
    }

    public void method8() {
        Integer m = 10; // violation 'Distance .* is 4.'
        class BClass extends Parent{
            public BClass getNewInstance(){
                return new BClass(){
                    private final int t = m;

                    @Override
                    void mm(){
                        System.out.println();
                    }
                };
            }
        }
    }
}
```

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.3-all.jar -c config.xml Test3.java
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-X.XX-all.jar -c config.xml Test3.java
```

```
Starting audit...
[WARN] /var/tmp/Test3.java:8:9: Distance between variable 'm' declaration and its first usage is 4, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [Custom]
Audit done.
```
---

**Describe what you expect in detail.**
No error is expected because the distance here is 1.

---

**Comment by jxr98** [comment]
I will work on it after #12822.

**Comment by Z4ID-Reaper** [comment]
I have read check documentation: https://checkstyle.org/config_coding.html#VariableDeclarationUsageDistance
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

`config.xml`
same as provided by the user above.
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
          <property name="id" value="Custom"/>
          <property name="validateBetweenScopes" value="true"/>
          <property name="allowedDistance" value="1"/>
        </module> 
    </module>
    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>
```
`Test.java`
this file is same as the one provided in https://github.com/checkstyle/checkstyle/issues/13011#issuecomment-1515484550 here

```
public class Test3 {
    class Parent {
        void mm() {
        }
    }

    public void method8() {
        Integer m = 10; // violation 'Distance .* is 4.'
        class BClass extends Parent{
            public BClass getNewInstance(){
                return new BClass(){
                    private final int t = m;

                    @Override
                    void mm(){
                        System.out.println();
                    }
                };
            }
        }
    }
}
```

`cli `

```
D:\zaid\zaid\codes\asfd>java -jar checkstyle-10.23.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```


i believe this issue was solved in the later versions of checkstyle.
hence making this issue invalid 

to test the correctness one more time i did the same config.xml file with a same Test.java   but a slight change of adding two print statements.

`Test.java`

```
public class Test3 {
    class Parent {
        void mm() {
        }
    }

    public void method8() {
        Integer m = 10; // violation 'Distance .* is 4.'
        System.out.println("Starting calculation...");
        System.out.println("Iterating...");
        class BClass extends Parent{
            public BClass getNewInstance(){
                return new BClass(){
                    private final int t = m;

                    @Override
                    void mm(){
                        System.out.println();
                    }
                };
            }
        }
    }
}
```

`cli`

```
D:\zaid\zaid\codes\asfd>java -jar checkstyle-10.23.0-all.jar -c config.xml Test.java
Starting audit...
[WARN] D:\zaid\zaid\codes\asfd\Test.java:8:9: Distance between variable 'm' declaration and its first usage is 3, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [Custom]
Audit done.
```

** ⚠ The above error comes up which was expected**

**Comment by Zopsss** [comment]
Can you test all of the remaining examples to make sure everything is working fine?

**Comment by Z4ID-Reaper** [comment]
i have read the documentation https://checkstyle.org/config_coding.html#VariableDeclarationUsageDistance
@Zopsss Appreciate the foresight .

i did find an unrelated error i will try to make it as readable as i can .
things to keep in mind:
1) i have used the same config file that was used by the reporter of the issue.

### config.xml 

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
          <property name="id" value="Custom"/>
          <property name="validateBetweenScopes" value="true"/>
          <property name="allowedDistance" value="1"/>
        </module> 
    </module>
    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>
```
---
### Individual Test files
```
public class Test {

    void nothing() {
    }

    class Parent {
        void mm() {
        }
    }

public void method2() {
        Integer a = 5;  // violation 'Distance .* is 4.'
        class AClass extends Parent {
            int i;

            public AClass(int i) {
                this.i = i;
            }

            @Override
            void mm() {
                if (a >= 0) {
                    System.out.println("test");
                }
            }
        }
    }
}
```

```
public class Test1 {
    public interface Ani {
        public void method4(Object o);
    }

    public void method5() {
        // false positive, the distance is 1
        Object a = new Object(); // violation 'Distance .* is 4.'
        class AClass implements Ani {
            @Override
            public void method4(Object o) {
                if (a.toString().isEmpty()) {
                    System.out.println(o.toString());
                }
            }
        }
    }
}
```

```
public class Test2 {
    public void method7() {
        Integer t = 5;
        nothing();
        System.out.println();
        class BClass extends Parent {
            @Override
            void mm() {
                System.out.println(t);
            }
        }
    }
}
```

```
public class Test3 {
    class Parent {
        void mm() {
        }
    }

    public void method8() {
        Integer m = 10; // violation 'Distance .* is 4.'
        class BClass extends Parent{
            public BClass getNewInstance(){
                return new BClass(){
                    private final int t = m;

                    @Override
                    void mm(){
                        System.out.println();
                    }
                };
            }
        }
    }
}
```

---
### cli output 

```
D:\zaid\zaid\codes\asfd>java -jar checkstyle-10.23.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.

D:\zaid\zaid\codes\asfd>java -jar checkstyle-10.23.0-all.jar -c config.xml Test1.java
Starting audit...
Audit done.

D:\zaid\zaid\codes\asfd>java -jar checkstyle-10.23.0-all.jar -c config.xml Test2.java
Starting audit...
Audit done.

D:\zaid\zaid\codes\asfd>java -jar checkstyle-10.23.0-all.jar -c config.xml Test3.java
Starting audit...
Audit done.

```

**Comment by Z4ID-Reaper** [comment]
@romani @Zopsss there is some more problems should i create a new issue to discuss this?

issues with when we switch between

1) <property name="validateBetweenScopes" value="false"/>
2) <property name="validateBetweenScopes" value="true"/>

**Comment by romani** [comment]
Please create new issue to show by CLI outputs nuances

**Comment by romani** [comment]
Please create new issue to show by CLI outputs nuances

**Comment by romani** [comment]
Reopening to as we have reference in code 
https://github.com/checkstyle/checkstyle/blob/1b76caf23ba759db39646c710410717918662e30/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceGeneral3.java#L91

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/0a341c4dc53b149ee42c5ddf8e5507d44738c6eb/src/main/resources/google_checks.xml#L320C19-L320C51

defalut is "false" for validateBetweenScopes
https://checkstyle.sourceforge.io/checks/coding/variabledeclarationusagedistance.html#Properties

so it is not Google style related issue

**Comment by Atharv3221** [comment]
Is anyone working on this ?

**Comment by romani** [comment]
nobody, please help us fix it.

**Comment by lithops-zty** [comment]
I'm on it.

Currently the following case seems to be erroneous. @romani I would like to confirm if the expect behavior I wrote matches the consensus.

#### Config File
`cat config.xml`
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
          <property name="id" value="Custom"/>
          <property name="validateBetweenScopes" value="true"/>
          <property name="allowedDistance" value="1"/>
        </module> 
    </module>
    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>
```

#### The Test
`cat Test.java`
```java
/*
VariableDeclarationUsageDistance
allowedDistance = 1
ignoreVariablePattern = (default)
validateBetweenScopes = true
ignoreFinal = false


*/

package com.puppycrawl.tools.checkstyle.checks.coding.variabledeclarationusagedistance;

import java.util.ArrayList;
import java.util.List;

public class Test {

    void nothing() {
    }

    class Parent {
        void mm() {
        }
        <T> void xx(List<T> m){}
    }

    public void method7() {
        // until https://github.com/checkstyle/checkstyle/issues/13011, the below distance is 3
        Integer t = 5;
        nothing();
        System.out.println();
        class BClass extends Parent {
            @Override
            void mm() {
                System.out.println(t);
            }
        }
    }
}

```

**Compilability Proof**
`javac Test.java`
```
```

**Actual Behavior**
`java -Duser.language=en -Duser.country=US -jar checkstyle-13.3.0-all.jar -c config.xml Test.java`
```
Starting audit...
Audit done.
```

**Expected Behavior**
`java -Duser.language=en -Duser.country=US -jar checkstyle-13.3.0-all.jar -c config.xml Test.java`
```
Starting audit...
[WARN] /mnt/d/IntellijProjects/checkstyle/Test.java:29:9: Distance between variable 't' declaration and its first usage is 3, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [Custom]
Audit done.
```
---
Despite the calculated distance being 3, the warning is exempted by the method https://github.com/checkstyle/checkstyle/blob/01b5c2ad215d0494867b72ca8aea0cb5ce84ad07/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L224, causing no warnings emitted eventually.

A revamping of this method is required if the test above is indeed a bug.

**Comment by romani** [comment]
orignal issue is resolve by some other update.

```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
          <property name="id" value="Custom"/>
          <property name="validateBetweenScopes" value="true"/>
          <property name="allowedDistance" value="1"/>
        </module> 
    </module>
    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>

$ cat Test.java 
public class Test {

    void nothing() {
    }

    class Parent {
        void mm() {
        }
    }

public void method2() {
        Integer a = 5;  // violation 'Distance .* is 4.'
        class AClass extends Parent {
            int i;

            public AClass(int i) {
                this.i = i;
            }

            @Override
            void mm() {
                if (a >= 0) {
                    System.out.println("test");
                }
            }
        }
    }
}

$ java -jar checkstyle-13.3.0-all.jar -c config.xml Test.java 
Starting audit...
Audit done.

```

**Comment by romani** [comment]
@lithops-zty, I do not think there should be improvements.

if more statement happening in between, distance counting good.
```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
          <property name="id" value="Custom"/>
          <property name="validateBetweenScopes" value="true"/>
          <property name="allowedDistance" value="1"/>
        </module> 
    </module>
    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>


$ cat Test.java 
public class Test {

    void nothing() {
    }

    class Parent {
        void mm() {
        }
    }

public void method2() {
        Integer a = 5; 
        nothing();
        class AClass extends Parent {
            int i;

            public AClass(int i) {
                this.i = i;
            }

            @Override
            void mm() {
                if (a >= 0) {
                    System.out.println("test");
                }
            }
        }
    }
}

$ java -jar checkstyle-13.3.0-all.jar -c config.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:12:9: Distance between variable 'a' declaration and its first usage is 2, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [Custom]
Audit done.

```

**Comment by romani** [comment]
@lithops-zty , you case is moved to https://github.com/checkstyle/checkstyle/issues/19156, as this issue becomes too big.

**Comment by lithops-zty** [comment]
@romani noted. The example you gave worked as expected because method `mm` has an `if` statement; false negative occurs when  `mm()` has only a single `println`. Will investigate further in the new issue.

**Comment by romani** [comment]
Please create new issue
